modules/launcher/src/test/java/org/mule/module/launcher/MuleApplicationClassLoaderTestCase.java

```diff
@@ -98,6 +102,32 @@ public void cleanUp()
     }
 
     @Test
+    public void leakCleanerIsCreatedByCorrectClassLoaderApp() throws Exception
+    {
+        ensureLeakCleanerIsCreatedByCorrectClassLoader(appCL);
+    }
+
+    @Test
+    public void leakCleanerIsCreatedByCorrectClassLoaderDomain() throws Exception
+    {
+        ensureLeakCleanerIsCreatedByCorrectClassLoader(domainCL);
+    }
+
+    private void ensureLeakCleanerIsCreatedByCorrectClassLoader(AbstractArtifactClassLoader artifactClassLoader) throws Exception
+    {
+        assertThat(artifactClassLoader.getClass().getClassLoader(), is(Thread.currentThread().getContextClassLoader()));
+        artifactClassLoader.setLeakCleanerClassName("TestLeakCleaner.class");
+        artifactClassLoader.dispose();
+
+        // We must call the getClassLoaderClassName method from TestLeakCleaner dynamically in order to not load the
+        // class by the current class loader if not a java.lang.LinkageError is raised.
+        Method getClassLoaderMethod = artifactClassLoader.getLeakCleanerInstance().getClass().getMethod("getClassLoader");
+        ClassLoader classLoader = (ClassLoader) getClassLoaderMethod.invoke(artifactClassLoader.getLeakCleanerInstance());
```

Comment:
Instead of defining getLeakCleanerInstance in the AbstractArtifactClasLoader, you can create a subclass of it and put that method there and use that instance on the test